src/go/overlord/statemanager.go

```diff
@@ -0,0 +1,168 @@
+	// AddInterest adds a new interest for a given minion to the state of the scan.
+	AddInterest(scanID string, interest *mpb.Interest, minion string) error
+	// CreateScan initializes the state for a scan.
+	CreateScan(scanID string) error
+	// GetFiles returns all the files known for a given ScanID
+	GetFiles(scanID string) ([]*pb.File, error)
+	// GetInterests returns all the interests known for a given ScanID, mapped to minions
+	GetInterests(scanID string) ([]*mappedInterest, error)
+	// RemoveFile atomically removes a given file from the state.
+	RemoveFile(scanID string, file *pb.File) error
+	// ScanExists returns true if any state at all is known about the scan.
+	ScanExists(scanID string) bool
+}
+
+// LocalStateManager handles state through a local time-expiring cache.
+type LocalStateManager struct {
+	lc *cache.Cache
+}
+
+// state stores the current state of a scan
+type state struct {
+	interests []*mappedInterest
+	files     map[string]*pb.File
+}
+
+// NewLocalStateManager creates a StateManager backed by a local cache.
+func NewLocalStateManager() *LocalStateManager {
+	lc := cache.New(5*time.Minute, 10*time.Minute)
+	return &LocalStateManager{lc: lc}
+}
+
+// AddFiles adds a set of files to the state. This will also dynamically
+// merge chunks of files.
+func (l *LocalStateManager) AddFiles(scanID string, files []*pb.File) error {
+	s, found := l.getState(scanID)
+	if !found {
+		return errors.New("Cannot find state of scan")
+	}
+	for _, f := range files {
+		currentFile, alreadyKnown := s.files[f.GetMetadata().GetPath()]
+		if !alreadyKnown {
+			currentFile = &pb.File{
+				Metadata: f.GetMetadata(),
+				DataChunks: []*pb.DataChunk{
+					&pb.DataChunk{
+						Offset: 0,
+					},
+				},
+			}
+			s.files[f.GetMetadata().GetPath()] = currentFile
+		}
+		currentChunk := currentFile.GetDataChunks()[0]
+		size := int64(len(currentChunk.GetData()))
+
+		newChunks := f.GetDataChunks()
+		sort.Slice(newChunks, func(i, j int) bool {
+			return newChunks[i].GetOffset() < newChunks[j].GetOffset()
+		})
+		for _, chunk := range newChunks {
+			if chunk.GetOffset() < size {
+				return fmt.Errorf("received a file with overlapping DataChunks")
+			}
+			if chunk.GetOffset() != size {
+				return fmt.Errorf("received a file with missing DataChunks")
+			}
+			currentChunk.Data = append(currentChunk.Data, chunk.GetData()...)
+			size += int64(len(chunk.GetData()))
+		}
+	}
+	l.setState(scanID, s)
+	return nil
+}
+
+// CreateScan initializes the state for a scan. It resets the state
+// if it already exists.
+func (l *LocalStateManager) CreateScan(scanID string) error {
+	l.setState(scanID, state{
+		interests: make([]*mappedInterest, 0),
+		files:     make(map[string]*pb.File),
+	})
+	return nil
+}
+
+// RemoveFile removes a given file from the state for a scan, if present.
+func (l *LocalStateManager) RemoveFile(scanID string, file *pb.File) error {
+	//l.lc.Delete(l.fileKey(scanID, file))
+	// TODO: implement me
+	return nil
```

Comment:
I guess return an error here

Reply:
No OK this is awful, let's fix it instead (see new commit)